### ISSUE TITLE
Template Triangle and Cicero Directory structure

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/cicero#435]
It would be nice if we could align the Cicero directory structure so it matches the conceptual description of an Accord Project template: https://docs.accordproject.org/docs/accordproject.html

The current directory structure (in `0.13.5`) is as follows:
```
package.json
    Metadata for the template
README.md
    A markdown file that describes the template

grammar/template.tem
    The default grammar for the template

models/
    Concerto model files for the template. They define the Template Model and models for the State, Request, Response, and Obligations used during execution.

lib/
    Ergo files specify the template business logic

sample.txt
    A sample clause or contract text
request.json
    A sample valid request for the template

test/
    A collection of unit tests for the template
```

We could make a few changes to the Accord Project "Template Triangle" and to the directory structure so they align. The resulting directory structure could become (here shown with changes related to `0.20.0` as well):
```
package.json
    Metadata for the template
README.md
    A markdown file that describes the template

grammar/template.tem.md
    The default grammar for the template

model/
    Concerto model files for the template. They define the Template Model and models for the State, Request, Response, and Obligations used during execution.

logic/
    Ergo files specify the template business logic

sample.md
    A sample clause or contract text
request.json
    A sample valid request for the template

test/
    A collection of unit tests for the template
```

This would facilitate documentation and first users could more quickly relate the conceptual description to the implementation in Cicero.

**Comment by jeromesimeon** [comment]
Corresponding change to the Accord Project Template Triangle could be as follows:
<img width="520" alt="Screenshot 2019-10-10 at 12 14 12 PM" src="https://user-images.githubusercontent.com/670099/66586982-8e160280-eb57-11e9-9bcc-e030649a44ad.png">

**Comment by jeromesimeon** [comment]
For info, here is the proposal’s Changelog:
1. On figure:
- `Natural Language` --> `Grammar`
- Flip `Grammar` and `Logic` and the corresponding arrows (suggested by @Katie Sa)

2. On Cicero directory:
- `sample.txt` --> `sample.md`
- `./grammar/template.tem` --> `./grammar/template.tem.md`
- `./lib` --> `./logic`
- `./models` --> `./model`

**Comment by irmerk** [comment]
I may have more thoughts in the future, just wanting to include @michael-grover on making the design cleaner.

**Comment by mttrbrts** [comment]
I vote for `logic`, `model` and `text`

**Comment by Michael-Grover** [comment]
![image](https://user-images.githubusercontent.com/40569793/66957430-39312b00-f034-11e9-8177-34f4b1304749.png)

![image](https://user-images.githubusercontent.com/40569793/66957450-451ced00-f034-11e9-8b35-dbc1ec387a76.png)

Some ideas, feedback welcome

**Comment by irmerk** [comment]
@Michael-Grover this last one looks the best to me.

**Comment by jeromesimeon** [comment]
> I vote for `logic`, `model` and `text`

Text confuses me a bit because I often make a distinction between `clause grammar` (the `.tem` or `.tem.md` file) and `clause text` (the `.txt` or `.md` file).

I would vote for `logic`, `model` and `language`.

I would be fine with all options (`text` / `language` / `grammar`)

**Comment by jeromesimeon** [comment]
> > I vote for `logic`, `model` and `text`
> 
> Text confuses me a bit because I often make a distinction between `clause grammar` (the `.tem` or `.tem.md` file) and `clause text` (the `.txt` or `.md` file).
> 
> I would vote for `logic`, `model` and `language`.
> 
> I would be fine with all options (`text` / `language` / `grammar`)

Addendum: 'Text` does look really good on the last figure @Michael-Grover posted.

**Comment by irmerk** [comment]
> Text confuses me a bit because I often make a distinction between `clause grammar` (the `.tem` or `.tem.md` file) and `clause text` (the `.txt` or `.md` file).

Good point. I vote for either `language` or `text`.

**Comment by mttrbrts** [comment]
@jeromesimeon Interesting. I see both clause grammar and clause text as belonging in the bottom-left bucket. One is parameterized text, the other is a textual instance of the template

**Comment by Michael-Grover** [comment]
![AP diagram-01](https://user-images.githubusercontent.com/40569793/66959577-a8a91980-f038-11e9-952c-878bae27d7f4.png)

**Comment by jeromesimeon** [comment]
> @jeromesimeon Interesting. I see both clause grammar and clause text as belonging in the bottom-left bucket. One is parameterized text, the other is a textual instance of the template

@mttrbrts Interesting. That interpretation would make me lean toward placing the `sample.md` files in the `./text` directory. But either way, I think I am getting quite on board with `text`.

I would love to know how people would feel about the file names for the grammar itself? Would something like this make sense:
```
./text/grammar.tem.md
./text/grammar_en.tem.md
./text/grammar_fr.tem.md
```
?

**Comment by dselman** [comment]
I’ve also see “verbalisation” used for this sort of thing, though that’s a bit of a mouthful.

I like the idea of moving sample.txt/MD under the “text” folder, as it too, should be per locale.

I’ve also been considering having templates for more than just the contract model; for example a template for contract state would allow us to generate a text representation for display. A template for request/response could also be useful.

How about:

/text/[en-US]/
... contract-model.tem.md
... contract-model.md (“sample”)
... contract-state.tem.md

Etc.

**Comment by jeromesimeon** [comment]
@dselman I really like the idea of having templates for other models (e.g., state/request/response) rather than just the contract or clause model, but it all feels a little too much to absorb for this release.

I feel there is quite a lot of support for using `./text` as the name in the triangle, so I would like to switch from `./grammar` to `./text` in the code and the documentation ASAP.

In addition to the already proposed directory reorganisation, would the following compromise for `0.20` be acceptable:
1. Rename the `grammar` directory to `text`
2. Move `sample*.md` files to the `text` directory

The corresponding new directory structure would look as follows:
```
The resulting directory structure could become (here shown with changes related to `0.20.0` as well):
```
package.json
    Metadata for the template
README.md
    A markdown file that describes the template

text/grammar.tem.md
    The contract or clause grammar for the template
text/sample.md or text/sample.md 
    A sample clause or contract text

model/
    Concerto model files for the template. They define the Template Model and models for the State, Request, Response, and Obligations used during execution.

logic/
    Ergo files specify the template business logic

request.json
    A sample valid request for the template

test/
    A collection of unit tests for the template
```

**Comment by irmerk** [comment]
@jeromesimeon I think this makes sense to me, I vote for it.

**Comment by Michael-Grover** [comment]
Does this require any changes to the graphic?

**Comment by jeromesimeon** [comment]
@Michael-Grover Short answer:no. Details: This was only adjusting the template directory to match your graphic.

I personally love the graphic and it's already used in the documentation: https://docs.accordproject.org/docs/next/accordproject-concepts.html

My only question is whether it would be useful to have a variant of it without the project names (Cicero/Concerto, etc) on it. This could be decided based on exposition needs (in documentation, in slide for live tutorials, etc).

**Comment by dselman** [comment]
> My only question is whether it would be useful to have a variant of it without the project names (Cicero/Concerto, etc) on it. This could be decided based on exposition needs (in documentation, in slide for live tutorials, etc).

On t-shirts! ;-)

**Comment by Michael-Grover** [comment]
@dselman  I'd love Accord Project logo T-shirts first :P . Peter told me we are waiting for funding from the Linux Foundation

**Comment by jeromesimeon** [comment]
> @dselman I really like the idea of having templates for other models (e.g., state/request/response) rather than just the contract or clause model, but it all feels a little too much to absorb for this release.
> 
> I feel there is quite a lot of support for using `./text` as the name in the triangle, so I would like to switch from `./grammar` to `./text` in the code and the documentation ASAP.
> 
> In addition to the already proposed directory reorganisation, would the following compromise for `0.20` be acceptable:
> 
> 1. Rename the `grammar` directory to `text`
> 2. Move `sample*.md` files to the `text` directory
> 3. Rename `template.tem.md` to something fairly generic e.g., `default.tem.md` or `grammar.tem.md`, so once we add verbalisation is will become clear that this is the default when no verbalisation is specified?
> 
> The corresponding new directory structure would look as follows:
> 
> ```
> package.json
>     Metadata for the template
> README.md
>     A markdown file that describes the template
> 
> text/grammar.tem.md
>     The contract or clause grammar for the template
> text/sample.md or text/sample.md 
>     A sample clause or contract text
> 
> model/
>     Concerto model files for the template. They define the Template Model and models for the State, Request, Response, and Obligations used during execution.
> 
> logic/
>     Ergo files specify the template business logic
> 
> request.json
>     A sample valid request for the template
> 
> test/
>     A collection of unit tests for the template
> ```

- This change has been implemented and is released in Cicero `0.20.0-beta.2`
- The proposed Diagram from @Michael-Grover aligns with that directory structure and is part of the documentation (for the upcoming released).

I'm closing this issue as resolved.

**Comment by jeromesimeon** [comment]
> I’ve also see “verbalisation” used for this sort of thing, though that’s a bit of a mouthful.
> 
> I like the idea of moving sample.txt/MD under the “text” folder, as it too, should be per locale.
> 
> I’ve also been considering having templates for more than just the contract model; for example a template for contract state would allow us to generate a text representation for display. A template for request/response could also be useful.
> 
> How about:
> 
> /text/[en-US]/
> ... contract-model.tem.md
> ... contract-model.md (“sample”)
> ... contract-state.tem.md
> 
> Etc.

This should be moved to #174
